x-pack/plugins/cloud_security_posture/server/routes/csp_rule_template/update_csp_rules_states.ts

```diff
@@ -0,0 +1,149 @@
+import type { CspRuleBulkActionRequest } from '../../../common/types';
+import type { CspRulesStates, CspSettings } from '../../../common/schemas/csp_settings';
+import { CspRouter } from '../../types';
+
+const muteStatesMap = {
+  mute: true,
+  unmute: false,
+};
+
+export const updateRulesStates = async (
+  soClient: SavedObjectsClientContract,
+  newRulesStates: CspRulesStates
+) => {
+  return await soClient.update<CspSettings>(
+    INTERNAL_CSP_SETTINGS_SAVED_OBJECT_TYPE,
+    INTERNAL_CSP_SETTINGS_SAVED_OBJECT_ID,
+    { rules_states: newRulesStates }
+  );
+};
+
+export const setRulesStates = (
+  rulesStates: CspRulesStates,
+  ruleIds: string[],
+  state: boolean
+): CspRulesStates => {
+  ruleIds.forEach((ruleId) => {
+    if (rulesStates[ruleId]) {
+      // Rule exists, set entry
+      rulesStates[ruleId] = { muted: state };
+    } else {
+      // Rule does not exist, create an entry
+      rulesStates[ruleId] = { muted: state };
+    }
+  });
+  return rulesStates;
+};
+
+export const createCspSettingObject = async (soClient: SavedObjectsClientContract) => {
+  return soClient.create<CspSettings>(
+    INTERNAL_CSP_SETTINGS_SAVED_OBJECT_TYPE,
+    {
+      rules_states: {},
+    },
+    { id: INTERNAL_CSP_SETTINGS_SAVED_OBJECT_ID }
+  );
+};
+
+export const getCspSettings = async (
+  soClient: SavedObjectsClientContract,
+  logger: Logger
+): Promise<CspSettings | undefined> => {
+  try {
+    const cspSettings = await soClient.get<CspSettings>(
+      INTERNAL_CSP_SETTINGS_SAVED_OBJECT_TYPE,
+      INTERNAL_CSP_SETTINGS_SAVED_OBJECT_ID
+    );
+    return cspSettings.attributes;
+  } catch (err) {
+    const error = transformError(err);
+    if (error.statusCode === 404) {
+      return undefined;
+    } else {
+      logger.error(`An error occurred while trying to fetch csp settings: ${error}`);
+    }
+  }
+};
+
+export const getCspSettingObjectSafe = async (
+  soClient: SavedObjectsClientContract,
+  logger: Logger
+): Promise<CspSettings> => {
+  const cspSettings = await getCspSettings(soClient, logger);
+
+  if (!cspSettings) return (await createCspSettingObject(soClient)).attributes;
+
+  return cspSettings;
+};
+
+export const defineUpdateCspRuleStateRoute = (router: CspRouter) =>
+  router.versioned
+    .post({
+      access: 'internal',
+      path: UPDATE_CSP_RULE_STATE_ROUTE_PATH,
+    })
+    .addVersion(
+      {
+        version: '1',
+        validate: {
+          request: {
+            query: cspRuleBulkActionRequest,
+          },
+        },
+      },
+      async (context, request, response) => {
+        if (!(await context.fleet).authz.fleet.all) {
+          return response.forbidden();
+        }
+        const cspContext = await context.csp;
+
+        try {
+          const requestBody: CspRuleBulkActionRequest = request.query;
+
+          const cspSettings = await getCspSettingObjectSafe(cspContext.soClient, cspContext.logger);
+
+          const currentRulesStates = cspSettings.rules_states;
```

Comment:
@CohenIdo partial update means that you can update specific fields of an object. In our use-case here, it means you don't need to fetch the entire object to make an update. You just send the update request and declare which fields you wish to update.

This will simplify your logic here :)